### PR TITLE
API: Disable API authentication

### PR DESCRIPTION
**Pull request by shiqizng**
This PR adds a new config option `DisableAPIAuth`. When set to true, public API endpoints are accessible without providing an auth token. And `DisableAPIAuth` is set to true for development profile. 

## Test
Started a container locally using development profile and verified that endpoints work without providing an API auth token.

### CODE REVIEW

File: cmd/algocfg/profileCommand.go [hunk 1]
```diff
@@ -44,6 +44,7 @@ var (
 			cfg.EnableDeveloperAPI = true
 			cfg.MaxAcctLookback = 256
 			cfg.EnableTxnEvalTracer = true
+			cfg.DisableAPIAuth = true
```
Comment (tzaffi) [inline cmd/algocfg/profileCommand.go:47]: It makes sense that with this new feature that the final version would be to default to disabling authentication for DeveloperAPI algod's. However, we need to communicate this change very loudly to the community. Unfortunately, there could be some people not following best practices and who block access to their developer endpoint via the token exclusively. If they don't look at the changelogs and upgrade their algod, they would be blindsided.

@algoanne

Comment (winder) [reply cmd/algocfg/profileCommand.go:47]: I'm not so concerned with the true defaults these days, but we could set it for the "development" profile in `algocfg`

Comment (winder) [reply cmd/algocfg/profileCommand.go:47]: lol, thats what you're commenting on! Good idea Shiqi!

File: config/localTemplate.go [hunk 2]
```diff
@@ -532,6 +532,8 @@ type Local struct {
 	// The private key provided must be an ed25519 private key.
 	// This is only used when P2PEnable is true. If the parameter is not set, it uses the default location.
 	P2PPrivateKeyLocation string `version[29]:""`
+
+	DisableAPIAuth bool `version[30]:"false"`
```
Comment (winder) [inline config/localTemplate.go:536]: Please add a comment here.

Comment (shiqizng) [reply config/localTemplate.go:537]: updated

**Comment by winder** [comment]
@shiqizng looks like there is a merge conflict

**Review by winder — DISMISSED** [review]

File: daemon/algod/server.go [hunk 3]
```diff
@@ -286,10 +286,15 @@ func (s *Server) Start() {
...
-	if err != nil {
-		fmt.Printf("APIToken error: %v\n", err)
-		os.Exit(1)
+	var apiToken string
+	var err error
+	fmt.Printf("API authentication disabled: %v\n", cfg.DisableAPIAuth)
+	if !cfg.DisableAPIAuth {
+		apiToken, err = tokens.GetAndValidateAPIToken(s.RootPath, tokens.AlgodTokenFilename)
+		if err != nil {
+			fmt.Printf("APIToken error: %v\n", err)
```
Comment (bbroder-algo) [inline daemon/algod/server.go:295]: maybe this error message could be enhanced by differentiating it from the same message on line 302 below which has a different token filename

File: daemon/algod/server.go [hunk 4]
```diff
@@ -286,10 +286,15 @@ func (s *Server) Start() {
...
 	}
 
-	apiToken, err := tokens.GetAndValidateAPIToken(s.RootPath, tokens.AlgodTokenFilename)
-	if err != nil {
-		fmt.Printf("APIToken error: %v\n", err)
-		os.Exit(1)
+	var apiToken string
+	var err error
+	fmt.Printf("API authentication disabled: %v\n", cfg.DisableAPIAuth)
+	if !cfg.DisableAPIAuth {
```
Comment (bbroder-algo) [inline daemon/algod/server.go:292]: Do you want to do this check in `cmd/algod/main.go` as well?

Comment (shiqizng) [reply daemon/algod/server.go:292]: in main, it generates a token if it doesn't exist. I don't think we want to change this behavior.

File: test/e2e-go/restAPI/restClient_test.go [hunk 5]
```diff
@@ -3459,3 +3459,23 @@ int 1
...
+
+	testClient := localFixture.LibGoalClient
+
+	statusResponse, err := testClient.Status()
+	a.NoError(err)
+	a.NotEmpty(statusResponse)
+	statusResponse2, err := testClient.Status()
+	a.NoError(err)
+	a.NotEmpty(statusResponse2)
+	a.True(statusResponse2.LastRound >= statusResponse.LastRound)
```
[the remark this answers is not in the document]
Comment (shiqizng) [reply test/e2e-go/restAPI/restClient_test.go:3480]: I had to run a container to verify this. is there a way to assert this in the test?

File: daemon/algod/server.go (hunk 4 above)
Comment (bbroder-algo) [reply daemon/algod/server.go:292]: but it also checks if the provided token has an error or not, which would be weird if `DisableAPIAuth` were true

File: test/e2e-go/restAPI/restClient_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbroder-algo) [reply test/e2e-go/restAPI/restClient_test.go:3480]: maybe you could write something in `go-algorand/test/e2e-go/cli/goal/expect`?

[the remark this answers is not in the document]
Comment (shiqizng) [reply test/e2e-go/restAPI/restClient_test.go:3483]: updated this test to include API calls, [92abc10](https://github.com/algorand/go-algorand/pull/5625/commits/92abc10a629fba67af397eafc42c65fad876c744)

**Review by winder — APPROVED** [review]

**Review by bbroder-algo — APPROVED** [review]

**Pull request merged**